### PR TITLE
Ac/refac/app phase to routes

### PR DESCRIPTION
**Pull request by aviadch**
App used to change the component according to a phase variable who lives in the context. 

It got very messy and made me making a lot of weird functions to change and get the phase. using router made it much easier and readable.
instead of changing the phase we change the route and the component render itself automatically.

### CODE REVIEW

File: frontend/src/App.js [hunk 1]
```diff
@@ -1,14 +1,17 @@
...
 import MyProvider from "./infra/MyProvider.js";
 
 import MafiaApp from "./MafiaApp";
 
 const App = () => {
   return (
-    <MyProvider>
-      <MafiaApp />
-    </MyProvider>
+    <BrowserRouter>
```
Comment (AdiYElisha) [inline frontend/src/App.js:10]: this is exactly why you need the App component, nice

File: frontend/src/MafiaApp.js [hunk 2]
```diff
@@ -1,33 +1,18 @@
-import React, { useContext } from "react";
+import React from "react";
 
 import WelcomeScreen from "./screens/WelcomeScreen";
 import WaitingRoom from "./screens/WaitingRoom";
-import { MyContext } from "./infra/MyContext";
-import { PHASE } from "./shared_code/consts.js";
+import { Switch, Route } from "react-router-dom";
 
 const MafiaApp = () => {
```
Comment (AdiYElisha) [inline frontend/src/MafiaApp.js:7]: is this the convention? not something like AppRoutes?

File: frontend/src/infra/MyProvider.js [hunk 3]
```diff
@@ -22,14 +21,15 @@ const MyProvider = (props) => {
...
-
+  let history = useHistory();
   socket.on("NewPlayer", (data) => {
     console.log("New Player from socket");
     setState({ ...state, playerList: data.roomPlayers });
   });
 
   const onNewGame = () => {
     console.log("OnNewGame Pressed");
+
```
Comment (AdiYElisha) [inline frontend/src/infra/MyProvider.js:32]: why so many whitespaces

File: frontend/src/infra/MyProvider.js [hunk 4]
```diff
@@ -40,12 +40,15 @@ const MyProvider = (props) => {
...
-          phase: PHASE.WAITING_ROOM,
           playerId: playerId,
           currentRoom: roomID,
           roomCreationDate: creationDate,
         });
+
         console.log(state);
+      })
+      .catch((err) => {
+        window.alert(`Server giving us something wrong! ${err}`);
```
Comment (AdiYElisha) [inline frontend/src/infra/MyProvider.js:51]: nice

File: frontend/src/MafiaApp.js (hunk 2 above)
Comment (aviadch) [reply frontend/src/MafiaApp.js]: AppRoutes sounds much better
9499074

File: frontend/src/infra/MyProvider.js (hunk 3 above)
Comment (aviadch) [reply frontend/src/infra/MyProvider.js]: Sometimes whitespaces make it more readable

File: backend/lib/server_code/api/roomRoutes.js [hunk 5]
```diff
@@ -38,18 +38,20 @@ roomRouter.post('/create', async (req, res) => {
   }
 });
 
-roomRouter.get('/join', (req, res) => {
-  const { userID, roomID, playerName } = req.query;
-  const playerToAdd = { id: userID, name: playerName };
+roomRouter.get('/register', (req, res) => {
+  console.log(req.query);
```
Comment (AdiYElisha) [inline backend/lib/server_code/api/roomRoutes.js:42]: LOGS

File: backend/lib/server_code/api/roomRoutes.js [hunk 6]
```diff
@@ -38,18 +38,20 @@ roomRouter.post('/create', async (req, res) => {
...
     const [roomSocket, roomSocketPort] = roomsSockets[roomID];
-    roomSocket.emit('PlayerJoinedRoom', {
-      message: 'A new user has joined the room',
-      roomPlayers,
+    roomSocket.emit('PlayerRegisteredRoom', {
+      message: 'A new player has registered to the room',
+      roomPlayersList,
     });
-    res.send({ joinDate: res.date, roomPlayers, roomSocketPort });
+    res.send({ registerDate: res.date, roomPlayersList, roomSocketPort });
```
Comment (AdiYElisha) [inline backend/lib/server_code/api/roomRoutes.js:54]: i don't really think we need the date

Comment (AdiYElisha) [reply backend/lib/server_code/api/roomRoutes.js:54]: also, no need to send him the roomPlayersList as we have a socket with that info

File: frontend/src/infra/MyProvider.js [hunk 7]
```diff
@@ -1,143 +1,107 @@
...
-    const playerId = Shortid.generate();
-    setState({
-      ...state,
-      phase: PHASE.WAITING_ROOM,
-      playerId,
+  const [room, roomDispatch] = useReducer(roomReducer, roomInitialState);
+
+  const updatePlayerList = (data) => {
+    roomDispatch({
+      type: 'updatePlayerList',
```
Comment (AdiYElisha) [inline frontend/src/infra/MyProvider.js:26]: make this a const, and put it outside this file

File: frontend/src/infra/MyProvider.js [hunk 8]
```diff
@@ -1,143 +1,107 @@
...
+    if (room.socket) {
+      verifySocketListen(room.socket, 'PlayerRegisteredRoom', updatePlayerList);
+    }
+  }, [room.socket]);
+
+  const history = useHistory();
 
+  const onRoomCreated = () => {
+    history.push(`/${ROOM_ROUTES}`);
+    playerDispatch({ type: 'generateUniqueID' });
```
Comment (AdiYElisha) [inline frontend/src/infra/MyProvider.js:43]: do it for all the types

Comment (AdiYElisha) [inline frontend/src/infra/MyProvider.js:43]: why do you playerDispatch on room created? only on register user

File: backend/lib/server_code/api/roomRoutes.js [hunk 9]
```diff
@@ -7,7 +7,7 @@ let roomRouter = express.Router();
 
 // Mockups
 let roomCreated = false;
-let roomPlayers = [];
+let roomPlayersList = [];
```
[the remark this answers is not in the document]
Comment (aviadch) [reply backend/lib/server_code/api/roomRoutes.js]: 942e2c9..21d7976

File: backend/lib/server_code/api/roomRoutes.js (hunk 6 above)
Comment (aviadch) [reply backend/lib/server_code/api/roomRoutes.js:54]: yes we do, socket is just for updating, the sending of playerlist is for init.
but you are right about the date so i delete it.

Comment (aviadch) [reply backend/lib/server_code/api/roomRoutes.js]: fe277e0

File: frontend/src/infra/MyProvider.js (hunk 7 above)
Comment (aviadch) [reply frontend/src/infra/MyProvider.js]: fe277e0

Comment (aviadch) [reply frontend/src/infra/MyProvider.js]: 4f14963198d17ff381fe702ef67f4589a5e548a8

File: frontend/src/infra/MyProvider.js (hunk 8 above)
Comment (aviadch) [reply frontend/src/infra/MyProvider.js]: 4f14963198d17ff381fe702ef67f4589a5e548a8

Comment (aviadch) [reply frontend/src/infra/MyProvider.js]: no! the room should need who created it

**Pull request merged**